src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -93,47 +121,86 @@ void Bank::evaluate_and_allocate() {
+                NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                            "Tensor should be registered first!");
+                auto uid = iter_device_registered->second;
+                if (device_bank.storage[uid].tensor) {
+                    // Already allocated
+                    continue;
+                }
+
+                uids_to_allocated[uid] = {true, lt.eval_meta(), ov::Tensor()};
             }
-            dev_guard.unlock();
-
-            // Allocation and/or evaluation needed
-            // Evaluate concurrently, lock the device
-            // mutex only to update the device bank (& allocate on-device memory, if needed)
-            const auto& transformed_tensor = lt.eval();
+        }
 
-            dev_guard.lock();
-            if (device_for_alloc == "CPU") {
-                // No allocation needed
-                device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = transformed_tensor;
-                const_cast<LazyTensor&>(lt).detach();
-                return;
+        std::unique_lock guard(device_bank.mutex);
+        // Allocate memory sequentially - in order of UID
+        for (std::size_t i = 0; i < uids_to_allocated.size(); ++i) {
+            auto& allocated = uids_to_allocated[i];
+            if (!allocated.needs_alloc) {
+                continue;
             }
+            auto uid = i;
 
             ov::SoPtr<ov::ITensor> remote_tensor;
-            ov::Tensor allocated_tensor;
 
-            auto remote_ctx = m_core->get_default_context(device_for_alloc)._ptr;
             remote_tensor =
-                remote_ctx->create_host_tensor(transformed_tensor.get_element_type(), transformed_tensor.get_shape());
-            allocated_tensor = ov::make_tensor(remote_tensor);
-            device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = allocated_tensor;
-            dev_guard.unlock();  // Unlock the guard, map update is done - copy can continue in parallel
+                get_context(m_core, device_for_alloc)->create_host_tensor(allocated.meta.type, allocated.meta.shape);
+            uids_to_allocated.at(uid) = {true, allocated.meta, ov::make_tensor(remote_tensor)};
+        }
+        guard.unlock();
+
+        // Copy to allocated memory
+        // Evaluate concurrently, lock the device
+        // mutex only to update the device bank (& allocate on-device memory, if needed)
+        ov::parallel_for(uids_to_allocated.size(), [&](std::size_t idx) {
+            std::unique_lock dev_guard(device_bank.mutex);
+            auto& allocated = uids_to_allocated[idx];
+            if (!allocated.needs_alloc) {
+                return;
+            }
+            auto uid = idx;
+            dev_guard.unlock();
 
-            transformed_tensor.copy_to(allocated_tensor);
+            auto transformed = device_bank.storage.at(uid).lt.eval();
```

Comment:
I'm still not sure what we are protecting with `device_bank.mutex`. But it seems to me that if we need to lock on line 114-122 then we need to protect this line too.
The same as line 198 and 203.